File: VB/WebApplication/XpoGate.asmx.vb
Imports System.Web.Services
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.DB.Exceptions

Public Enum ServiceException
    None
    Schema
End Enum
' To allow this Web Service to be called from script, using ASP.NET AJAX, uncomment the following line. 
<WebService(Namespace:="http://tempuri.org/"), WebServiceBinding(ConformsTo:=WsiProfiles.BasicProfile1_1), System.Web.Script.Services.ScriptService()> _
Public Class XpoGate
    Inherits System.Web.Services.WebService
    Private Shared store As IDataStore
    Shared Sub New()
        Dim connStr As String = MSSqlConnectionProvider.GetConnectionString("(local)", "NorthwindXpo")
        store = XpoDefault.GetConnectionProvider(connStr, AutoCreateOption.SchemaAlreadyExists)
    End Sub
    <WebMethod()> _
    Public Function SelectData(ByVal selects() As SelectStatement, <System.Runtime.InteropServices.Out()> ByRef e As ServiceException) As SelectedData
        Try
            e = ServiceException.None
            Return store.SelectData(selects)
        Catch e1 As SchemaCorrectionNeededException
            e = ServiceException.Schema
            Return Nothing
        End Try
    End Function
    '<WebMethod()> _
    'Public Function GetAutoCreateOption() As AutoCreateOption
    '    Return AutoCreateOption.SchemaAlreadyExists
    'End Function
    '<WebMethod()> _
    'Public Function UpdateSchema(ByVal dontCreateIfFirstTableNotExist As Boolean, ByVal tables() As DBTable) As UpdateSchemaResult
    '    Return store.UpdateSchema(dontCreateIfFirstTableNotExist, tables)
    'End Function
    <WebMethod()> _
    Public Function ModifyData(ByVal statements() As ModificationStatement) As ModificationResult
        Return store.ModifyData(statements)
    End Function
End Class

File: CS/WebApplication/XpoGate.asmx.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Data;
using System.Linq;
using System.Web;
using System.Web.Services;
using System.Web.Services.Protocols;
using System.Xml;
using System.Xml.Serialization;
using DevExpress.Data.Filtering;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.DB.Exceptions;

namespace XpoWebService {
	public enum ServiceException { None, Schema }
    [WebService(Namespace = "http://tempuri.org/")]
	[WebServiceBinding(ConformsTo = WsiProfiles.BasicProfile1_1)]
	[ToolboxItem(false)]
	// To allow this Web Service to be called from script, using ASP.NET AJAX, uncomment the following line. 
	[System.Web.Script.Services.ScriptService]
	public class XpoGate : System.Web.Services.WebService {
		static IDataStore store;
		static XpoGate() {
            string connStr = MSSqlConnectionProvider.GetConnectionString("(local)", "NorthwindXpo");
			store = XpoDefault.GetConnectionProvider(connStr, AutoCreateOption.SchemaAlreadyExists);
		}
		[WebMethod]
		public SelectedData SelectData(SelectStatement[] selects, out ServiceException e) {
			try {
				e = ServiceException.None;
                return store.SelectData(selects);
			} catch(SchemaCorrectionNeededException) {
				e = ServiceException.Schema;
				return null;
			}
		}
        //[WebMethod]
        //public AutoCreateOption GetAutoCreateOption() {
        //    return AutoCreateOption.SchemaAlreadyExists;
        //}
        //[WebMethod]
        //public UpdateSchemaResult UpdateSchema(bool dontCreateIfFirstTableNotExist, DBTable[] tables) {
        //    return store.UpdateSchema(dontCreateIfFirstTableNotExist, tables);
        //}
		[WebMethod]
		public ModificationResult ModifyData(ModificationStatement[] statements) {
			return store.ModifyData(statements);
		}
	}
}